File: use-cases/01-simple-completion/python/main.py
"""
Polyglot Agent Labs — Use Case 01: Simple LLM Completion (Multi-Provider)
Sends a prompt to an LLM via litellm, supporting OpenAI, Anthropic, and OpenRouter.
Switch provider with env var LLM_PROVIDER (default: openrouter).
"""

import os
import sys

from dotenv import load_dotenv
from litellm import completion


# Provider → model mapping
PROVIDERS = {
    "openai": "openai/gpt-4.1-nano",
    "anthropic": "anthropic/claude-3-haiku-20240307",
    "openrouter": "openrouter/stepfun/step-3.5-flash:free",
}

PROMPT = "Hello! Tell me a fun fact about programming."


def main():
    load_dotenv()

    provider = os.getenv("LLM_PROVIDER", "openrouter").lower()

    if provider not in PROVIDERS:
        print(f"✗ Unknown provider: '{provider}'")
        print(f"  Supported: {', '.join(PROVIDERS.keys())}")
        sys.exit(1)

    model = PROVIDERS[provider]

    print("=== Python — Simple LLM Completion ===")
    print(f"Provider:  {provider}")
    print(f"Model:     {model}")
    print()

    try:
        response = completion(
            model=model,
            messages=[{"role": "user", "content": PROMPT}],
        )

        text = response.choices[0].message.content
        usage = response.usage

        print(f"Prompt:    {PROMPT}")
        print(f"Response:  {text}")
        print()
        if usage:
            print(f"Tokens:    {usage.prompt_tokens} in / {usage.completion_tokens} out / {usage.total_tokens} total")

    except Exception as e:
        print(f"✗ Error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: use-cases/01-simple-completion/rust/src/main.rs
//! Polyglot Agent Labs — Use Case 01: Simple LLM Completion (Multi-Provider)
//! Sends a prompt to an LLM via rig-core, supporting OpenAI, Anthropic, and OpenRouter.
//! Switch provider with env var LLM_PROVIDER (default: openrouter).

use std::env;

use anyhow::{bail, Result};
use rig::agent::AgentBuilder;
use rig::client::CompletionClient;
use rig::completion::Prompt;
use rig::providers::{anthropic, openai, openrouter};

const PROMPT_TEXT: &str = "Hello! Tell me a fun fact about programming.";

#[tokio::main]
async fn main() -> Result<()> {
    let provider = env::var("LLM_PROVIDER")
        .unwrap_or_else(|_| "openrouter".to_string())
        .to_lowercase();

    println!("=== Rust — Simple LLM Completion ===");
    println!("Provider:  {provider}");

    let response = match provider.as_str() {
        "openai" => {
            let model_id = "gpt-4.1-nano";
            println!("Model:     {model_id}");
            println!();
            let api_key = env::var("OPENAI_API_KEY")
                .map_err(|_| anyhow::anyhow!("OPENAI_API_KEY not set"))?;
            let client = openai::Client::new(&api_key)?;
            let model = client.completion_model(model_id);
            let agent = AgentBuilder::new(model).build();
            agent.prompt(PROMPT_TEXT).await?
        }
        "anthropic" => {
            let model_id = "claude-3-haiku-20240307";
            println!("Model:     {model_id}");
            println!();
            let api_key = env::var("ANTHROPIC_API_KEY")
                .map_err(|_| anyhow::anyhow!("ANTHROPIC_API_KEY not set"))?;
            let client = anthropic::Client::new(&api_key)?;
            let model = client.completion_model(model_id);
            let agent = AgentBuilder::new(model).build();
            agent.prompt(PROMPT_TEXT).await?
        }
        "openrouter" => {
            let model_id = "stepfun/step-3.5-flash:free";
            println!("Model:     {model_id}");
            println!();
            let api_key = env::var("OPENROUTER_API_KEY")
                .map_err(|_| anyhow::anyhow!("OPENROUTER_API_KEY not set"))?;
            let client = openrouter::Client::new(&api_key)?;
            let model = client.completion_model(model_id);
            let agent = AgentBuilder::new(model).build();
            agent.prompt(PROMPT_TEXT).await?
        }
        _ => {
            bail!("Unknown provider: '{provider}'. Supported: openai, anthropic, openrouter");
        }
    };

    println!("Prompt:    {PROMPT_TEXT}");
    println!("Response:  {response}");

    Ok(())
}
